lib/ymodel/loadable.rb

```diff
@@ -0,0 +1,44 @@
+# frozen_string_literal: true
+
+require 'ymodel/schema'
+require 'ymodel/errors'
+
+module YModel
+  # Module resposible for loading the schema and data from the yaml files.
+  module Loadable
+    # This method can be called from within a concrete implementation to
+    # overwrite the default ymodel filename associated with that model.
+    def source_file(filename)
+      @source = filename
+
+      # Similar to YModel::Trigger#inherited. A hook for loading the schema.
+      define_readers self
+    end
+
+    def schema
+      YModel::Schema.new(records)
+    rescue Errno::ENOENT
+      YModel::Schema.new({})
+    end
+
+    protected
+
+    def records
```

Comment:
I think the performance gain would be marginal.

Reply:
I am thoroughly dissapointed in my ability to express my thoughts in English, Here's a shot at responding at your feedback though!

YModel isn't finished and I'm not expecting it will be "finished" before we merge this. I'm not really considering what the "final" state of YModel should be. I'm trying to make the migration easy to understand.
I've prioritized ease of modifying the library itself and transparency/revertability of the migration over "robustness", ease of use or improving the naming and topology of the "interface" parts of the application. These things can (and should) be improved at a later time.

The PR has come a long way since the last review. Nora helped me with migrating Slides and InputElements and I've fixed some of the issues you pointed out.

Migrated models so far:
- Tab
- SidebarItem
- Slide
- InputElement

Bugfixes/Improvements:
- Keyword arguments removed from #where and #initialize
- YModel::Base#where uses a param whitelist before calling send
- Use string interpolation in YModel::Bas#initialize instead of concatenation
- Fixed the data in YModel::Schema. Kind of weird to store the sourcefile there.

extra features:
  - Implemented find_by
  - Implemented attribute?

I'm planning to do a couple of other things before the merge. Most notably do something about the area dependencies in the migrated models and playing nice with the linter. I'm not to keen on adding an attributes API or using DrySchema for now. My reasons are explained below.

Some other issues you brought up that I'd like to address in isolation:

#### A transparent data migration
> Data from the DB has been dumped into YAML without any changes that I can see. I'd like to understand why... was it a conscious decision, or simply due to lack-of-time to make improvements?

The reason I didn't mutate any data is to keep the migration transparent. It makes it a bit simpler to see the "diff" when not considering its structure.
I'm only trying solve a single problem with this mr, I know it got a bit convoluted, my apologies! Some things had to change though: I broke tests, relation methods and views on the way and removed everything I thought was junk.

#### Nested objects in YAML
> `parent_id` could be removed by allowing models to be nested.

Might be a good idea. I think it might require a bit more time than I have at the moment. Writing something to test referential integrity should be a lot simpler. I think it might be more useful as well since it would be very reusable.

#### Migrate Area dependencies
> Sidebar items should be able to be hidden depending on whether an attribute in the area data (retrieved from ETEngine) is true or false. This is purpose of the AreaDependency model and area_dependency relation. It seems no longer possible for users to add/edit/delete such dependencies for Tab and SidebarItem.

I think AreaDependencies should be moved into the migrated models as attributes. I was a bit reluctant to do that in light of a "transparent and reversible" migration but I'm pretty sure its best to get it over with. I'll add it before we merge.

#### YModel data environment isolation
> * It would be nice to have proper test isolation from the data in config/ymodel. The tests shouldn't be dependent on real-world data. I suspect this is why you hit problems with the input element controller spec.

I don't fully agree with you here. I'll argue this data is a part of the 'Behaviour' of the application more then it is part of its 'Environment'.
Modifying YModel to source from a different folder in between environments would be simple enough but I'm afraid that it would make testing less realistic. I'm of the opinion a controller spec shouldn't be that familiar with the contents the underlying models anyways. I'm curious wether you agree my line of reasoning.

#### Explicit attribute declaration
> * I'm a big fan of explicitly naming the attributes used by a class, rather than inferring them magically from the data like ActiveRecord does. I think AR gets away with it because the DB schema and schema.rb keeps things strict.

Explicitly naming the attributes isn't a terrible idea but I've opted to not add an attribute API yet. I am aware that we lose some safety for now but its a trade-off I'm willing to make for additional flexibility at this moment. Besides that, I think we might end up with only two "types" of data in the YModels. Strings and relations. 

There are also other ways to improve safety and legibility. I think writing specs per model or a YModel precondition combined with an annotate clone would be nicer and much more in line with Ruby than an attributes API build using dry.rb libraries.

Some options that come to mind are
- an Annotate clone.
- a YModel precondition to check the YAML files for consistency.
- an attribute API.
- Written specs per model.

Would you mind if I postpone adding an attribute api?